File: .vscode/jira-proxy.py
#!/usr/bin/env python3
import ssl, http.server, http.client, socketserver

TARGET_HOST = 'issue.swf.i.mercedes-benz.com'
LISTEN_PORT = 18080
CERT = '/lhome/giatran/.ssh/daimler/client.pem'
KEY  = '/lhome/giatran/.ssh/daimler/client.key'
CA   = '/etc/ssl/certs/ca-certificates.crt'

class ProxyHandler(http.server.BaseHTTPRequestHandler):
    def proxy(self):
        ctx = ssl.create_default_context(cafile=CA)
        ctx.load_cert_chain(certfile=CERT, keyfile=KEY)
        conn = http.client.HTTPSConnection(TARGET_HOST, 443, context=ctx)
        length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(length) if length else None
        hdrs = {k: v for k, v in self.headers.items()
                if k.lower() not in ('host', 'connection', 'transfer-encoding')}
        try:
            conn.request(self.command, self.path, body=body, headers=hdrs)
            r = conn.getresponse()
            data = r.read()
            self.send_response(r.status)
            for k, v in r.getheaders():
                if k.lower() not in ('transfer-encoding', 'connection'):
                    self.send_header(k, v)
            self.send_header('Content-Length', str(len(data)))
            self.end_headers()
            self.wfile.write(data)
            self.wfile.flush()
        except Exception as e:
            self.send_error(502, str(e))
        finally:
            conn.close()
    def log_message(self, fmt, *a): print(f'[jira-proxy] {fmt % a}')
    do_GET = do_POST = do_PUT = do_DELETE = do_PATCH = proxy

if __name__ == '__main__':
    class ThreadedServer(socketserver.ThreadingMixIn, http.server.HTTPServer):
        daemon_threads = True
    s = ThreadedServer(('127.0.0.1', LISTEN_PORT), ProxyHandler)
    print(f'Jira mTLS proxy on http://localhost:{LISTEN_PORT} -> {TARGET_HOST}', flush=True)
    s.serve_forever()

File: .vscode/jira-proxy.mjs
// Local mTLS proxy for issue.swf.i.mercedes-benz.com
// Listens on localhost:18080, forwards to Jira with client cert

import http from 'http';
import https from 'https';
import fs from 'fs';
import { URL } from 'url';

const TARGET = 'https://issue.swf.i.mercedes-benz.com'\;
const PORT = 18080;
const CERT_FILE = '/lhome/giatran/.ssh/daimler/client.pem';

const pemData = fs.readFileSync(CERT_FILE, 'utf8');
const certMatch = pemData.match(/-----BEGIN CERTIFICATE-----[\s\S]+?-----END CERTIFICATE-----/);
const keyMatch = pemData.match(/-----BEGIN PRIVATE KEY-----[\s\S]+?-----END PRIVATE KEY-----/);

const agent = new https.Agent({
  cert: certMatch[0],
  key: keyMatch[0],
  ca: fs.readFileSync('/etc/ssl/certs/ca-certificates.crt'),
});

const server = http.createServer((req, res) => {
  const targetUrl = new URL(TARGET + req.url);
  const options = {
    hostname: targetUrl.hostname,
    port: 443,
    path: targetUrl.pathname + targetUrl.search,
    method: req.method,
    headers: { ...req.headers, host: targetUrl.hostname },
    agent,
  };

  const proxyReq = https.request(options, (proxyRes) => {
    res.writeHead(proxyRes.statusCode, proxyRes.headers);
    proxyRes.pipe(res);
  });

  proxyReq.on('error', (e) => {
    console.error('Proxy error:', e.message);
    res.writeHead(502);
    res.end(e.message);
  });

  req.pipe(proxyReq);
});

server.listen(PORT, '127.0.0.1', () => {
  console.log(`Jira mTLS proxy running on http://localhost:${PORT}`);
});
